<launch>

    <arg name="world_name" default="garage"/>

    <arg name="vehicleHeight" default="0.75"/>
    <arg name="cameraOffsetZ" default="0"/>
    <arg name="vehicleX" default="0"/>
    <arg name="vehicleY" default="0"/>
    <arg name="terrainZ" default="0"/>
    <arg name="vehicleYaw" default="0"/>
    

    <arg name="robot1_namespace" default="robot_1"/>
    <arg name="robot2_namespace" default="robot_2"/>
    <arg name="robot3_namespace" default="robot_3"/>

    <arg name="checkTerrainConn" default="false"/>


    <!-- 启动 gazebo -->
    <include file="$(find gazebo_ros)/launch/empty_world.launch" >
        <arg name="gui" value="true" />
        <arg name="use_sim_time" value="true"/>
        <arg name="world_name" value="$(find p2os_urdf)/world/$(arg world_name).world"/>
    </include>

    <!-- Launch the first robot (robot1) with its namespace -->
    <group ns="$(arg robot1_namespace)">
        <!-- 将 Urdf 文件的内容加载到参数服务器 -->
        <param name="robot_description" command="$(find xacro)/xacro $(find p2os_urdf)/defs/pioneer3at.xacro robot_namespace:=$(arg robot1_namespace)" />
        <!-- 向rviz发布机器人link和model state中的tf变换 -->
        <!-- 关节以及机器人状态发布节点 -->
        <node name="joint_state_publisher" pkg="joint_state_publisher" type="joint_state_publisher" />
        <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher" >
            <param name="tf_prefix" type="string" value="$(arg robot1_namespace)" />
        </node>
        <!-- 在 gazebo 中显示机器人模型 -->
        <node pkg="gazebo_ros" type="spawn_model" name="model" args="-urdf -model $(arg robot1_namespace) -param robot_description -x 1"  >
            <!-- <param name="robot_namespace" type="string" value="$(arg robot1_namespace)"/> -->
        </node>

        <include file="$(find vehicle_simulator)/launch/pioneer_simulator.launch" >
            <arg name="name_space" value="$(arg robot1_namespace)"/>
            <arg name="vehicleHeight" value="$(arg vehicleHeight)"/>
            <arg name="cameraOffsetZ" value="$(arg cameraOffsetZ)"/>
            <arg name="vehicleX" value="$(arg vehicleX)"/>
            <arg name="vehicleY" value="$(arg vehicleY)"/>
            <arg name="terrainZ" value="$(arg terrainZ)"/>
            <arg name="vehicleYaw" value="$(arg vehicleYaw)"/>
        </include>
    
        <include file="$(find local_planner)/launch/local_planner.launch" >
            <arg name="name_space" value="$(arg robot1_namespace)"/>
            <arg name="cameraOffsetZ" value="$(arg cameraOffsetZ)"/>
            <arg name="goalX" value="$(arg vehicleX)"/>
            <arg name="goalY" value="$(arg vehicleY)"/>
        </include>
    
        <include file="$(find terrain_analysis)/launch/terrain_analysis.launch" />
    
        <include file="$(find terrain_analysis_ext)/launch/terrain_analysis_ext.launch" >
          <arg name="checkTerrainConn" value="$(arg checkTerrainConn)"/>
        </include>

        <include file="$(find ps3joy)/launch/ps3.launch" />

        <node launch-prefix="nice" pkg="rviz" type="rviz" name="rvizGA" args="-d $(find vehicle_simulator)/rviz/vehicle_simulator.rviz" respawn="true"/>
    </group>



    <node pkg="tf2_ros" type="static_transform_publisher" name="static_tf_pub_map_to_odom" args="0 0 0 0 0 0 map odom" />
  
    <!-- <include file="$(find sensor_scan_generation)/launch/sensor_scan_generation.launch" /> -->
  
  </launch>
  